### PR TITLE
*: let TempIndex support encode/decode partitionID flag

### LINKED ISSUE #56535: unexpected assertion failed using global index
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

Here is the log from our internal test schrddl. 
```
2024/10/09 16:48:19 ddl.go:212: [0;31m[fatal] [error] [instance 16] ERROR: Error 8141: assertion failed: key: 7480000000000000ba5f69800000000000000307800029e4698b0200, assertion: Exist, start_ts: 453113916671131764, existing start ts: 0, existing commit ts: 0
/build/ddl/dml_ops.go:52: Error when executing SQL: delete from `b5d472af` where not( IsNull( `b5d472af`.`56b0971e` ) ) or `b5d472af`.`56b0971e` in ( 1444517 ,824880 ) 
```

### 2. What did you expect to see? (Required)
No error
### 3. What did you see instead (Required)
assertion failed
### 4. What is your TiDB version? (Required)
master

### PR DESCRIPTION
**Pull request by Defined2014**
### What problem does this PR solve?

Issue Number: close #56535

Problem Summary: TempIndex didn't process partition handle before. So let it supports encode/decode partition handle.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by tangenta — APPROVED** [review]

File: pkg/tablecodec/tablecodec.go [hunk 1]
```diff
@@ -1415,6 +1438,16 @@ func (v *TempIndexValueElem) DecodeOne(b []byte) (remain []byte, err error) {
 			v.Handle, _ = kv.NewCommonHandle(b[:hLen])
 		}
 		b = b[hLen:]
+		if b[0] == TempIndexKeyTypePartitionIDFlag {
```
Comment (mjonss) [inline pkg/tablecodec/tablecodec.go:1441]: Is this only needed here, and not also in `case TempIndexValueFlagNormal:`?

**Review by mjonss — COMMENTED** [review]
Can you explain the issue a bit more technical? Is it only on delete (part of update) that the TempIndex would fail?

File: pkg/kv/key.go [hunk 2]
```diff
@@ -683,9 +683,9 @@ func (m *MemAwareHandleMap[V]) Range(fn func(h Handle, val V) bool) {
 			return
```
[the remark this answers is not in the document]
Comment (Defined2014) [reply pkg/kv/key.go:683]: Could both used, but I think it's not happened in TiDB.

File: pkg/tablecodec/tablecodec.go [hunk 3]
```diff
@@ -1307,16 +1307,31 @@ func (v TempIndexValue) FilterOverwritten() TempIndexValue {
...
+//   - [flag 1 byte][handle_length 2 bytes] [handle handle_len bytes] [partitionIdFlag 1 byte] [partitionID 8 bytes] [key_version 1 byte] {distinct deleted}
 //   - [flag 1 byte]                                                  [key_version 1 byte] {non-distinct deleted}
 type TempIndexValueElem struct {
 	Value    []byte
 	Handle   kv.Handle
 	KeyVer   byte
 	Delete   bool
 	Distinct bool
+
+	Global bool
```
Comment (Defined2014) [inline pkg/tablecodec/tablecodec.go:1319]:
```suggestion
    // Global means it's a global Index, for partitioned tables. Only for `distinct` + `deleted` scenario right now.
	Global bool
```

File: pkg/tablecodec/tablecodec.go (hunk 1 above)
Comment (Defined2014) [reply pkg/tablecodec/tablecodec.go:1442]: Yes, only delete will include handle. https://github.com/pingcap/tidb/pull/57017/files#diff-d38a794eb5c0ced0fe2e07bc0af6bd8624e360f4c5055b6c7075d62afac48569R1310

Comment (Defined2014) [reply pkg/tablecodec/tablecodec.go:1442]: Also https://github.com/pingcap/tidb/pull/57017/files#diff-d38a794eb5c0ced0fe2e07bc0af6bd8624e360f4c5055b6c7075d62afac48569L1338

File: pkg/tablecodec/tablecodec.go [hunk 4]
```diff
@@ -1307,16 +1307,32 @@ func (v TempIndexValue) FilterOverwritten() TempIndexValue {
...
+//   - [flag 1 byte][handle_length 2 bytes] [handle handle_len bytes] [partitionIdFlag 1 byte] [partitionID 8 bytes] [key_version 1 byte] {distinct deleted}
 //   - [flag 1 byte]                                                  [key_version 1 byte] {non-distinct deleted}
 type TempIndexValueElem struct {
 	Value    []byte
 	Handle   kv.Handle
 	KeyVer   byte
 	Delete   bool
 	Distinct bool
+
+	// Global means it's a global Index, for partitioned tables. Only for `distinct` + `deleted` scenario right now.
```
Comment (Defined2014) [inline pkg/tablecodec/tablecodec.go:1319]:
```suggestion
	// Global means it's a global Index, for partitioned tables. Currently only used in `distinct` + `deleted` scenarios.
```

**Review by mjonss — APPROVED** [review]

**Pull request merged**